File: ApolloLLM/app/src/main/java/com/apollo/MainActivity.java
package com.apollo;

import android.app.Activity;
import android.os.Bundle;
import android.widget.TextView;
import android.widget.LinearLayout;
import android.widget.Button;
import android.content.ContentValues;
import android.database.sqlite.SQLiteDatabase;

public class MainActivity extends Activity {

    static {
        System.loadLibrary("apollo");
    }

    public native String runWorkflowTransition(String state, String operator);
    public native String getWorkflowGraph();
    public native void rollback();

    private TextView statusView;
    private DatabaseHelper dbHelper;

    @Override
    protected void onCreate(Bundle b) {
        super.onCreate(b);
        dbHelper = new DatabaseHelper(this);

        LinearLayout layout = new LinearLayout(this);
        layout.setOrientation(LinearLayout.VERTICAL);

        statusView = new TextView(this);
        statusView.setText("Ready.");
        layout.addView(statusView);

        Button btnStep = new Button(this);
        btnStep.setText("Execute Step (O_PHYS)");
        btnStep.setOnClickListener(v -> {
            String newState = runWorkflowTransition("current", "O_PHYS");
            persistTransition("O_PHYS", newState);
            updateUI();
        });
        layout.addView(btnStep);

        Button btnRollback = new Button(this);
        btnRollback.setText("Rollback");
        btnRollback.setOnClickListener(v -> {
            rollback();
            updateUI();
        });
        layout.addView(btnRollback);

        setContentView(layout);
        updateUI();
    }

    private void persistTransition(String op, String state) {
        SQLiteDatabase db = dbHelper.getWritableDatabase();
        ContentValues values = new ContentValues();
        values.put("operator", op);
        values.put("state", state);
        db.insert("workflow_state", null, values);
    }

    private void updateUI() {
        statusView.setText(getWorkflowGraph());
    }
}

File: ApolloLLM/app/src/main/java/com/apollo/MainActivity.kt
package com.apollo

import android.app.Activity
import android.os.Bundle
import android.widget.TextView
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.Button
import android.widget.EditText
import android.app.AlertDialog

class MainActivity : Activity() {

    companion object {
        init {
            System.loadLibrary("apollo")
        }
    }

    external fun runWorkflowTransition(state: String, operator: String): String
    external fun getWorkflowGraph(): String
    external fun rollback()

    private lateinit var statusView: TextView

    override fun onCreate(b: Bundle?) {
        super.onCreate(b)

        val layout = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
        }

        val header = TextView(this).apply {
            text = "Chronofold_X: Sovereign DAG IDE"
            textSize = 20f
            setPadding(20, 20, 20, 20)
        }
        layout.addView(header)

        val btnLayout = LinearLayout(this).apply {
            orientation = LinearLayout.HORIZONTAL
        }

        Button(this).apply {
            text = "Add Node"
            setOnClickListener { showAddNodeDialog() }
        }.also { btnLayout.addView(it) }

        Button(this).apply {
            text = "Import YAML"
            setOnClickListener { /* TODO */ }
        }.also { btnLayout.addView(it) }

        Button(this).apply {
            text = "Build APK"
            setOnClickListener { /* TODO */ }
        }.also { btnLayout.addView(it) }

        layout.addView(btnLayout)

        statusView = TextView(this).apply {
            text = "Ready."
            setPadding(20, 20, 20, 20)
        }

        val scroll = ScrollView(this).apply {
            addView(statusView)
        }
        layout.addView(scroll, LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT,
            0, 1f
        ))

        setContentView(layout)
        updateUI()
    }

    private fun showAddNodeDialog() {
        val input = EditText(this)
        AlertDialog.Builder(this)
            .setTitle("Insert Node")
            .setView(input)
            .setPositiveButton("OK") { _, _ ->
                runWorkflowTransition("current", "ADD_NODE_${input.text}")
                updateUI()
            }
            .setNegativeButton("Cancel", null)
            .show()
    }

    private fun updateUI() {
        statusView.text = getWorkflowGraph()
    }
}
